hal/spi_api.h

```diff
@@ -19,220 +18,305 @@
- *      maximum of tx_length and rx_length.
- */
-int spi_master_block_write(spi_t *obj, const char *tx_buffer, int tx_length, char *rx_buffer, int rx_length, char write_fill);
-
-/** Check if a value is available to read
+ * @param[in,out] obj   A pointer to a spi_t object.
+ * @param[in]     init  Initialization parameters.
  *
- * @param[in] obj The SPI peripheral to check
- * @return non-zero if a value is available
+ * @return SPI_RESULT_OK on success. See spi_result_t for more details about failures.
  */
-int  spi_slave_receive(spi_t *obj);
+spi_result_t spi_init(spi_t *obj, spi_init_t *init);
 
-/** Get a received value out of the SPI receive buffer in slave mode
+/**
+ * Processes a transfer blocking until completion.
+ * This function locks the peripheral and prevents any other transaction to occur until the
+ * operation is completed (including asynchronous transactions).
+ * It will return `true` on success or `false` if :
+ * - at least one of its parameters is undefined (NULL).
+ * - the peripheral is busy with another transaction (from this spi_t or any other sharing the same peripheral).
+ * - any error occurs during the transfer. Errors can be but are not limited to :
+ *   - conflicting master ;
+ *   - rx or tx buffer overflow ;
+ *   - rx or tx buffer underflow.
  *
- * Blocks until a value is available
- * @param[in] obj The SPI peripheral to read
- * @return The value received
- */
-int  spi_slave_read(spi_t *obj);
-
-/** Write a value to the SPI peripheral in slave mode
+ * It sends `tx_count` symbols from the buffer pointed by `tx` or a fill symbol is `tx` is NULL.
+ * It receives `rx_count` symbols from the peripheral and stores them to `rx` if it is not NULL, else
+ * it will discard the read symbols.
  *
- * Blocks until the SPI peripheral can be written to
- * @param[in] obj   The SPI peripheral to write
- * @param[in] value The value to write
- */
-void spi_slave_write(spi_t *obj, int value);
-
-/** Checks if the specified SPI peripheral is in use
+ * @param[in,out] obj   A pointer to a spi_t object.
+ * @param[in,out] args  A pointer to a spi_transfer_args_t object.
  *
- * @param[in] obj The SPI peripheral to check
- * @return non-zero if the peripheral is currently transmitting
+ * @return True on success.
  */
-int  spi_busy(spi_t *obj);
+bool spi_transfer(spi_t *obj, spi_transfer_args_t *args);
 
-/** Get the module number
+/**
+ * Schedules a transfer using the given parameters.
+ * @warning All parameters and the embedded references must stay "alive" until completion of the operation.
+ *
+ * @param[in,out] obj       A pointer to a spi_t object.
+ * @param[in,out] args      A pointer to a spi_transfer_args_t object.
+ * @param[in]     context   A context to be used by the callback.
+ * @param[in]     cb        A callback invoked upon completion of this transaction.
  *
- * @param[in] obj The SPI peripheral to check
- * @return The module number
+ * @return A reference counted handle to the transaction.
```

Comment:
What is this returning? `spi_async_handle_t` is specified as being implementation specific , why ? what is expected to be defined there? each async transfer generates unique handle?

What if SPI peripheral is already taken - already transfer registered (SPI_0 peripheral already in use)? What this returns?

Reply:
where the peripheral is in use or not, this method creates and schedule a transaction to be processed as the peripheral becomes available.

This handle is used by the `_abort` method. If the user does not want to abort he can `spi_async_free_handle()` right away.
The resources attached to the handle (passed in `spi_async_tranfer()`) must stay allocated until completion of the transaction. This completion is notified by a call to the callback.

Reply:
if for some reason the transaction cannot be scheduled (queue is full etc) this returns NULL and does not take ownership of the data. The callback is therefore not invoked.

Reply:
I dont follow, not clear to me what `spi_async_handle_t ` is the meaning and why is it passed nor how transactions should work (I would assume transactions are handled in the upper layer, this one is just getting data and handles it)

Reply:
Hm, I see. I will rework that to make it clearer then.